pkg/monitor/controller.go

```diff
@@ -143,40 +198,169 @@ func (monitor *controllerMonitor) deleteStaleAgentCRDs() {
+		return true
+	} else if err := c.syncNode(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
+		// another change happens.
+		c.nodeQueue.Forget(key)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.nodeQueue.AddRateLimited(key)
+		klog.ErrorS(err, "Error syncing Node", "Node", key)
+	}
+	return true
+}
+
+func (c *controllerMonitor) processNextExternalNodeWorkItem() bool {
+	obj, quit := c.externalNodeQueue.Get()
+	if quit {
+		return false
+	}
+	defer c.externalNodeQueue.Done(obj)
+
+	if key, ok := obj.(string); !ok {
+		c.externalNodeQueue.Forget(obj)
+		klog.Errorf("Expected string in ExternalNode work queue but got %#v", obj)
+		return true
+	} else if err := c.syncExternalNode(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
+		// another change happens.
+		c.externalNodeQueue.Forget(key)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.externalNodeQueue.AddRateLimited(key)
+		klog.ErrorS(err, "Error syncing ExternalNode", "ExternalNode", key)
+	}
+	return true
+}
+
+func (c *controllerMonitor) syncNode(key string) error {
+	_, name, err := splitKeyFunc(key)
+	if err != nil {
+		// This err should not occur.
+		return err
+	}
+	_, err = c.nodeLister.Get(name)
+	if err != nil {
 		if errors.IsNotFound(err) {
-			monitor.deleteAgentCRD(crd.Name)
+			return c.deleteAgentCRD(name)
+		} else {
+			return err
 		}
 	}
+	return c.createAgentCRD(name)
+
 }
 
-func (monitor *controllerMonitor) deleteStaleAgentCRD(old interface{}) {
-	node, ok := old.(*corev1.Node)
-	if !ok {
-		tombstone, ok := old.(cache.DeletedFinalStateUnknown)
-		if !ok {
-			klog.Errorf("Error decoding object when deleting Node, invalid type: %v", old)
-			return
+func (c *controllerMonitor) syncExternalNode(key string) error {
+	namespace, name, err := splitKeyFunc(key)
+	if err != nil {
+		// This err should not occur.
+		return err
+	}
+	_, err = c.externalNodeLister.ExternalNodes(namespace).Get(name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return c.deleteAgentCRD(name)
+		} else {
+			return err
 		}
-		node, ok = tombstone.Obj.(*corev1.Node)
-		if !ok {
-			klog.Errorf("Error decoding object tombstone when deleting Node, invalid type: %v", tombstone.Obj)
-			return
+	}
+	return c.createAgentCRD(name)
+
+}
+
+func (monitor *controllerMonitor) createAgentCRD(name string) error {
+	klog.InfoS("Creating agent monitoring CRD", "name", name)
+	agentCRD := new(v1beta1.AntreaAgentInfo)
+	agentCRD.Name = name
+	_, err := monitor.client.CrdV1beta1().AntreaAgentInfos().Create(context.TODO(), agentCRD, metav1.CreateOptions{})
+	if err != nil {
+		if errors.IsAlreadyExists(err) {
+			klog.InfoS("Skipping creating agent monitoring CRD as it already exists", "name", name)
+		} else {
+			klog.ErrorS(err, "Failed to create agent monitoring CRD")
```

Comment:
We typically log in the caller. Could you check if no duplicated logs?

Reply:
Thanks, I have removed the lines to avoid the duplication here.